src/Features/Core/Portable/MakeMethodSynchronous/AbstractMakeMethodSynchronousCodeFixProvider.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MakeMethodSynchronous
+{
+    internal abstract class AbstractMakeMethodSynchronousCodeFixProvider : CodeFixProvider
+    {
+        public static readonly string EquivalenceKey = FeaturesResources.Make_method_synchronous;
+
+        protected abstract bool IsMethodOrAnonymousFunction(SyntaxNode node);
+        protected abstract SyntaxNode RemoveAsyncTokenAndFixReturnType(IMethodSymbol methodSymbolOpt, SyntaxNode node, ITypeSymbol taskType, ITypeSymbol taskOfTType);
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var document = context.Document;
+            var diagnostic = context.Diagnostics.First();
+
+            var token = diagnostic.Location.FindToken(cancellationToken);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var node = token.GetAncestor(IsMethodOrAnonymousFunction);
+            if (node != null)
+            {
+                context.RegisterCodeFix(new MyCodeAction(c => FixNodeAsync(context.Document, node, c)), diagnostic);
+            }
+        }
+
+        private const string AsyncSuffix = "Async";
+
+        private async Task<Solution> FixNodeAsync(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            // See if we're on an actual method declaration (otherwise we're on a lambda declaration).
+            // If we're on a method declaration, we'll get an IMethodSymbol back.  In that case, check
+            // if it has the 'Async' suffix, and remove that suffix if so.
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var methodSymbolOpt = semanticModel.GetDeclaredSymbol(node) as IMethodSymbol;
+
+            if (methodSymbolOpt?.MethodKind == MethodKind.Ordinary && 
+                methodSymbolOpt.Name.Length > AsyncSuffix.Length && 
+                methodSymbolOpt.Name.EndsWith(AsyncSuffix))
```

Comment:
No.  The convention is that you end an Async method with "Async".  Not "async" or "ASYNC" or anything like that.